### PR TITLE
KAFKA-3392: ConsumerRecords iterator throws NoSuchElementException when a TopicPartition is empty

### PR DESCRIPTION
**Pull request by drausin**
This contribution is my original work, and I license it under the project's open source license.

CC @jkreps

### CODE REVIEW

**Comment by guozhangwang** [comment]
cc @hachikuji , could we ever get the map of records with some of the partition containing empty set?

**Comment by drausin** [comment]
@guozhangwang, @hachikuji, what happens when we have fewer records than `TopicPartition`s? The empty `TopicPartitions` aren't in the map?

**Comment by drausin** [comment]
And if it's the case that we assume no partition has an empty set of records, perhaps it's worth asserting that in the constructor with something like a `Preconditions.checkArgument(!entry.getValue().isEmpty())` for each entry of `records`. I'd be happy to make that change if that's preferable.

**Comment by hachikuji** [comment]
@guozhangwang I think we currently only add to ConsumerRecords non-empty record sets (or this would have been found much sooner). That said, the change seems fine to me and protects anyone who construct ConsumerRecords instances in test cases where record sets may be empty.

@drausin Do we really need powermock to test this? Couldn't you just pass in a Map with a couple empty record lists and verify that the ConsumerRecords iterator behaves correctly?

**Comment by drausin** [comment]
@hachikuji, definitely don't need PowerMock, though I think it makes the test a tad cleaner, since we don't have to create a bunch of dummy instances.  That said, I'm happy to abide by the conventions/practices of the codebase, so I'll update the test per your suggestion.

**Comment by guozhangwang** [comment]
@hachikuji cool, I will leave it for you to review then.

File: clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerRecordsTest.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        ConsumerRecord<Integer, String> record1 = new ConsumerRecord<>(topic, 1, 0, 0L, TimestampType.CREATE_TIME, 0L, 0, 0, 1, "value1");
+        ConsumerRecord<Integer, String> record2 = new ConsumerRecord<>(topic, 1, 0, 0L, TimestampType.CREATE_TIME, 0L, 0, 0, 2, "value2");
+        records.put(new TopicPartition(topic, 1), Arrays.asList(record1, record2));
+
+        ConsumerRecords<Integer, String> consumerRecords = new ConsumerRecords<>(records);
+        Iterator<ConsumerRecord<Integer, String>> iter = consumerRecords.iterator();
+
+        int c = 0;
+        for (; iter.hasNext(); iter.next()) {
+            c++;
```
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerRecordsTest.java]: Would it be a good idea to make this test case a little stronger and assert something about the records (e.g. partition and offset)?

File: clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerRecordsTest.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    @Test
+    public void iterator() throws Exception {
+
+        Map<TopicPartition, List<ConsumerRecord<Integer, String>>> records = new HashMap<>();
+
+        String topic = "topic";
+        records.put(new TopicPartition(topic, 0), new ArrayList<ConsumerRecord<Integer, String>>());
+        ConsumerRecord<Integer, String> record1 = new ConsumerRecord<>(topic, 1, 0, 0L, TimestampType.CREATE_TIME, 0L, 0, 0, 1, "value1");
+        ConsumerRecord<Integer, String> record2 = new ConsumerRecord<>(topic, 1, 0, 0L, TimestampType.CREATE_TIME, 0L, 0, 0, 2, "value2");
+        records.put(new TopicPartition(topic, 1), Arrays.asList(record1, record2));
```
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerRecordsTest.java:43]: Would it be worthwhile to also add an empty collection at the end of the map as well to make sure that case is covered? You could use LinkedHashMap to ensure order.

**Comment by drausin** [comment]
@hachikuji, thanks for your comments. I've incorporated all of them into e448356. Please let me know if there's anything else you think would be helpful for me to add.

One additional question I didn't catch when reviewing the Kafka PR policies: do we squash PR commits into a single one before it gets merged into `trunk`?

**Comment by hachikuji** [comment]
@drausin Thanks for the update! I left a minor comment, but otherwise LGTM. Don't worry about squashing; our merge script does it automatically.

File: clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerRecordsTest.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        records.put(new TopicPartition(topic, 2), new ArrayList<ConsumerRecord<Integer, String>>());
+
+        ConsumerRecords<Integer, String> consumerRecords = new ConsumerRecords<>(records);
+        Iterator<ConsumerRecord<Integer, String>> iter = consumerRecords.iterator();
+
+        int c = 0;
+        for (;iter.hasNext(); c++) {
```
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerRecordsTest.java]: This line is failing checkstyle. I think we need a space after the first semicolon.

**Comment by hachikuji** [comment]
LGTM

**Comment by drausin** [comment]
Looks like I don't have the power to merge, so I'll leave it to you or the powers that be to handle that :)

**Comment by guozhangwang** [comment]
LGTM, merged to trunk.

**Comment by drausin** [comment]
Great. Many thanks @guozhangwang!

**Pull request closed without merging**